From ca51496b2cebf5b6df7514d53bfebe2acb039ef1 Mon Sep 17 00:00:00 2001
From: Erik Fleckstein <erik@tinkerforge.com>
Date: Wed, 16 Oct 2024 11:07:19 +0200
Subject: [PATCH] ETH: Do not enable DHCP by default.

Leaving DHCP enabled by default leads to race conditions during
address configuration. Disable it by default because it can be
turned on if actually required.

As this fixes the race condition, the 50ms delay to work around it
isn't required anymore.
---
 libraries/Ethernet/src/ETH.cpp | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/libraries/Ethernet/src/ETH.cpp b/libraries/Ethernet/src/ETH.cpp
index 13f9c11f..5b43e599 100644
--- a/libraries/Ethernet/src/ETH.cpp
+++ b/libraries/Ethernet/src/ETH.cpp
@@ -284,6 +284,9 @@ bool ETHClass::begin(eth_phy_type_t type, int32_t phy_addr, int mdc, int mdio, i
   esp_netif_config.if_key = if_key_str;
   esp_netif_config.if_desc = if_desc_str;
   esp_netif_config.route_prio -= _eth_index * 5;
+  // Do not enable DHCP by default.
+  // Also disable "GAP" while we're at it. It's not Gratuitous ARP but instead polls the device's own IP every minute.
+  esp_netif_config.flags = (esp_netif_flags_t)(esp_netif_config.flags - ESP_NETIF_DHCP_CLIENT - ESP_NETIF_FLAG_GARP);
 
   cfg.base = &esp_netif_config;
 
@@ -357,10 +360,6 @@ bool ETHClass::begin(eth_phy_type_t type, int32_t phy_addr, int mdc, int mdio, i
     }
   }
 
-  // holds a few milliseconds to let DHCP start and enter into a good state
-  // FIX ME -- addresses issue https://github.com/espressif/arduino-esp32/issues/5733
-  delay(50);
-
   return true;
 
 err:
-- 
2.47.0

